fl4health/losses/mkmmd_loss.py

```diff
@@ -77,156 +80,218 @@ def construct_quadruples(self, X: torch.Tensor, Y: torch.Tensor) -> torch.Tensor
+    def compute_euclidean_inner_products(self, X: torch.Tensor, Y: torch.Tensor) -> torch.Tensor:
+        # In this function, we assume that X, Y: n_samples, n_features
+        # We want to compute estimates of the expectation for each RBF kernel WITHOUT using a linear approximation.
+        # So we need to compute ||x - y||^2 for all pairs (x_j, x_k), (x_j, y_k), (x_k, y_j), and (y_j, y_k) for all
+        # j, k in range(n_samples).
+        # NOTE: ||x - y||^2 = <x - y, x - y> = <x, x> + <y, y> - 2<x, y>
         x_x_prime = (
             torch.sum((X**2), dim=1).reshape(-1, 1)
             + torch.sum((X**2), dim=1).reshape(1, -1)
             - 2.0 * torch.mm(X, torch.transpose(X, 0, 1))
         )
-        x_x_prime[x_x_prime < 0] = 0
         y_y_prime = (
             torch.sum((Y**2), dim=1).reshape(-1, 1)
             + torch.sum((Y**2), dim=1).reshape(1, -1)
             - 2.0 * torch.mm(Y, torch.transpose(Y, 0, 1))
         )
-        y_y_prime[y_y_prime < 0] = 0
         x_y_prime = (
             torch.sum((X**2), dim=1).reshape(-1, 1)
             + torch.sum((Y**2), dim=1).reshape(1, -1)
             - 2.0 * torch.mm(X, torch.transpose(Y, 0, 1))
         )
-        x_y_prime[x_y_prime < 0] = 0
         x_prime_y = (
             torch.sum((Y**2), dim=1).reshape(-1, 1)
             + torch.sum((X**2), dim=1).reshape(1, -1)
             - 2.0 * torch.mm(Y, torch.transpose(X, 0, 1))
         )
+
+        # Correct any values that ended up nearly but not identically zero (||x-y||^2 should be semi-definite)
+        x_x_prime[x_x_prime < 0] = 0
+        y_y_prime[y_y_prime < 0] = 0
+        x_y_prime[x_y_prime < 0] = 0
         x_prime_y[x_prime_y < 0] = 0
-        # each inner product is a tensor of dimension len(v_i_quadruples) x len(v_i_quadruples), we return a
+
+        # each inner product is a tensor of dimension n_samples x n_samples, we return a
         # tensor of shape 4 x len(X) x len(Y)
         return torch.cat(
-            [x_x_prime.unsqueeze(0), y_y_prime.unsqueeze(0), x_y_prime.unsqueeze(0), x_prime_y.unsqueeze(0)], dim=0
+            [x_x_prime.unsqueeze(0), y_y_prime.unsqueeze(0), x_y_prime.unsqueeze(0), x_prime_y.unsqueeze(0)]
         )
 
+    def compute_euclidean_inner_products_linear(self, v_i_quadruples: torch.Tensor) -> torch.Tensor:
+        # Shape of v_i_quadruples is n_samples/2 x 4 x n_features
+        # v_i = [x_{2i-1}, x_{2i}, y_{2i-1}, y_{2i}]
+        #
+        # We want to compute the RBF kernel values. To do this, we need to compute ||x - y||^2 for the relevant pairs
+        # x and y. That is the inner product. Note that ||x - y||^2 = <x - y, x - y> = (x-y)^T(x-y)
+        # For the quadruples of the form (x,  x', y, y') we need distances for pairings (x, x'), (y, y'), (x, y'),
+        # (x, y')
+        x_x_prime = torch.sum((v_i_quadruples[:, 0, :] - v_i_quadruples[:, 1, :]) ** 2, dim=1, keepdim=True)
```

Comment:
I went back to the torch.sum here rather than doing the quadratic expansion that you had. This is the "true" linear approximation suggested by Gretton. It doesn't seem to work 🤦, but we want to implement it faithfully. I removed the torch.mean from the original implementation because that was dumb of me. It should just be sum.

Reply:
Actually if I remember it correctly, later on I change sum to mean as the values were getting so big for QP computation. I am still not sure why it can mess up things as it is just a linear scaling to control effect of high number of dataset features.

Reply:
Ah you might be right. You're correct that it's just a weighted version of the inner product. So I think it's fine. However, since we offer the possibility of normalizing the features, I think we're okay to just leave it as a sum, unless you disagree 🙂

Reply:
let's keep it sum for now to be consistent with non-linear version.